geth/swarm.go

```diff
@@ -0,0 +1,186 @@
+package geth
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/contracts/chequebook"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm"
+	bzzapi "github.com/ethereum/go-ethereum/swarm/api"
+	"path/filepath"
+	"sync"
+)
+
+var (
+	ErrInvalidSwarmService           = errors.New("swarm service is unavailable")
+	ErrSwarmIdentityInjectionFailure = errors.New("failed to inject identity into Swarm")
+)
+
+type SwarmService struct {
+	swarm  *swarm.Swarm
+	keys   map[string]*ecdsa.PrivateKey
+	keysMu sync.RWMutex // Mutex to sync identity keys
+}
+
+func NewSwarmService() *SwarmService {
+	service := &SwarmService{
+		keys: make(map[string]*ecdsa.PrivateKey),
+	}
+
+	return service
+}
+
+// activateEthService configures and registers the eth.Ethereum service with a given node.
+func (self *SwarmService) StartSwarm(prvkey *ecdsa.PrivateKey) error {
+	if self.swarm != nil {
+		err := self.Stop()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	nm := NodeManagerInstance()
+	stack := nm.node.geth
+
+	chbookaddr := common.Address{}
+	bzzdir := filepath.Join(stack.InstanceDir(), "swarm")
+
+	networkId := uint64(1)
+	if UseTestnet {
+		networkId = 3
+	}
+
+	bzzconfig, err := bzzapi.NewConfig(bzzdir, chbookaddr, prvkey, networkId)
+	if err != nil {
+		return fmt.Errorf("%v: %v", err, err)
+	}
+
+	var client *ethclient.Client
+	client, err = ethclient.Dial(fmt.Sprintf("http://%s", stack.HTTPEndpoint()))
+	if err != nil {
+		return fmt.Errorf("Can't connect: %v", err)
+	}
+
+	self.swarm, err = swarm.NewSwarm(nil, client, bzzconfig, false, true, "")
+	if err != nil {
+		return err
+	}
+
+	return self.Start(nm.node.geth.Server())
+}
+
+func (self *SwarmService) Start(net *p2p.Server) error {
+	if self.swarm != nil && net != nil {
+		return self.swarm.Start(net)
+	}
+
+	return nil
+}
+
+func (self *SwarmService) Stop() error {
+	if self.swarm != nil {
+		err := self.swarm.Stop()
+		if err != nil {
+			return err
+		}
+
+		self.swarm = nil
+	}
+
+	return nil
+}
+
+func (self *SwarmService) Protocols() []p2p.Protocol {
```

Comment:
- purpose of introducing `Protocols()` and `APIs()` (below) methods?
- your custom `SwarmService`, that you introduced is supposed to be just a wrapper containing some config + `*swarm.Swarm`
- now, on why you've introduced these methods: to satisfy `node.Service` interface in https://github.com/status-im/status-go/pull/106/files#diff-dacdf432329ae450c554894d234b257dR225
- but then in addition to that you also needed to implement the whole `swarm_api_proxy.go`, again why?
- instead of producing all that code you could have rewrite swarm registration methods as:

```go
func activateSwarmService(stack *node.Node) error {
	serviceConstructor := func(ctx *node.ServiceContext) (node.Service, error) {
		swarmService := NewSwarmService()
		return swarmService.swarm, nil
	}
	if err := stack.Register(serviceConstructor); err != nil {
		return err
	}

	return nil
}
```

That's 2 lines of change! Note: you will need to init `*swarm.Swarm` in `NewSwarmService()` - even with randomly generated key, if necessary (basically, you need to call `StartSwarm(key)` in that constructor. 



To sum up: No need to introduce some crazy proxies and interfaces (which you don't need to satisfy semantically). Rely on `SwarmService.swarm` if you need to satisfy `node.Service` interface, no need to overextend interface to where it shouldn't belong.